proxy/src/proxy.rs

```diff
@@ -150,58 +159,74 @@ async fn handshake<S: AsyncRead + AsyncWrite + Unpin>(
-    let cancel_key_data = session.enable_cancellation(cancel_closure);
-
-    client
-        .write_message_noflush(&BeMessage::ParameterStatus(
-            BeParameterStatusMessage::ServerVersion(&version),
-        ))?
-        .write_message_noflush(&Be::BackendKeyData(cancel_key_data))?
-        .write_message(&BeMessage::ReadyForQuery)
-        .await?;
-
-    // This function will be called for writes to either direction.
-    fn inc_proxied(cnt: usize) {
-        // Consider inventing something more sophisticated
-        // if this ever becomes a bottleneck (cacheline bouncing).
-        NUM_BYTES_PROXIED_COUNTER.inc_by(cnt as u64);
+}
+
+impl<S> Client<S> {
+    /// Construct a new connection context.
+    fn new(stream: PqStream<S>, creds: auth::ClientCredentials) -> Self {
+        Client { stream, creds }
     }
+}
 
-    let mut db = MetricsStream::new(db, inc_proxied);
-    let mut client = MetricsStream::new(client.into_inner(), inc_proxied);
-    let _ = tokio::io::copy_bidirectional(&mut client, &mut db).await?;
+#[derive(Debug, Error)]
+enum Error {
+    #[error(transparent)]
+    Auth(#[from] auth::AuthError),
 
-    Ok(())
+    #[error(transparent)]
+    Compute(#[from] compute::ConnectionError),
 }
 
-/// Connect to a corresponding compute node.
-async fn connect_to_db(
-    db_info: DatabaseInfo,
-) -> anyhow::Result<(TcpStream, String, CancelClosure)> {
-    // TODO: establish a secure connection to the DB
-    let socket_addr = db_info.socket_addr()?;
-    let mut socket = TcpStream::connect(socket_addr).await?;
+impl<S: AsyncRead + AsyncWrite + Unpin> Client<S> {
+    /// Let the client authenticate and connect to the designated compute node.
+    async fn connect_to_db(
+        mut self,
+        config: &ProxyConfig,
+        session: cancellation::Session<'_>,
+    ) -> anyhow::Result<()> {
+        // Authenticate and connect to a compute node.
+        let connect = async {
+            let db_info = self.creds.authenticate(config, &mut self.stream).await?;
+            let (db, version, cancel_closure) = db_info.connect().await?;
+            let cancel_key_data = session.enable_cancellation(cancel_closure);
+
+            // `Error` type acts as an allowlist for errors.
+            Ok::<_, Error>((db, version, cancel_key_data))
+        };
 
-    let (client, conn) = tokio_postgres::Config::from(db_info)
-        .connect_raw(&mut socket, NoTls)
-        .await?;
+        let (db, version, cancel_key_data) = match connect.await {
+            Ok(x) => x,
+            Err(e) => return self.stream.throw_error(e).await,
+        };
+
+        self.stream
+            .write_message_noflush(&BeMessage::ParameterStatus(
+                BeParameterStatusMessage::ServerVersion(&version),
+            ))?
+            .write_message_noflush(&Be::BackendKeyData(cancel_key_data))?
+            .write_message(&BeMessage::ReadyForQuery)
+            .await?;
```

Comment:
Not new with this patch I guess, but why does the proxy need to create these messages? The Postgres server should send these right after authentication, doesn't the proxy forward them to the client?

Reply:
> Not new with this patch, but why does the proxy need to create these messages?

The reason for that is how we organize the authentication process. Basically, at the moment postgres takes no place in authentication. All clients are authenticated by the console (via an http method), and whatever messages get sent by compute node are consumed by the proxy. This isn't going to change when we move to SCRAM.